### PR TITLE
Merge feature/dataprovider to master

### LINKED ISSUE #8524: Merge feature/dataprovider to master
Merge the DataView API from the dataprovider feature branch to master.

Remember to remove component submodules from the feature before merging.

### PR DESCRIPTION
**Pull request by mshabarov**
Fixes #8524

### CODE REVIEW

**Comment by pleku** [comment]
@mshabarov this contains some changes that are already in master, why is that ?

**Comment by mshabarov** [comment]
> @mshabarov this contains some changes that are already in master, why is that ?

I've merged `master` to `feature/dataprovider` as to be confident there are no conflicts. Should I re-create the PR so as to exclude that merging? Code diff does not include the changed not related to dataprovider though.

**Comment by pleku** [comment]
>I've merged master to feature/dataprovider as to be confident there are no conflicts. Should I re-create the PR so as to exclude that merging? Code diff does not include the changes not related to dataprovider though.

Ok, I'm not sure either why those are there but I guess it is fine then.

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * @param <T>
+ *         data type
+ * @since
+ */
+public interface DataView<T> extends Serializable {
+
+    /**
+     * Get the full data available to the component.
+     * Data will use set filters and sorting.
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java:37]:
```suggestion
     * Data is filtered and sorted the same way as in the component.
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @return filtered and sorted data set
+     */
+    Stream<T> getItems();
+
+    /**
+     * Get the full data size with filters if any set.
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java:44]:
```suggestion
     * Gets the size of the data source with filters applied if any are set.
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @return filtered data size
+     */
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java:47]:
```suggestion
     * @see #addSizeChangeListener(ComponentEventListener<SizeChangeEvent<?>>)
     */
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @param listener
+     *         size change listener to register
+     * @return registration for removing the listener
+     */
+    Registration addSizeChangeListener(
+            ComponentEventListener<SizeChangeEvent<?>> listener);
+
+    /**
+     * Sets identity provider to be used for getting item identifier and
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java:85]:
```suggestion
     * Sets an identifier provider to be used for comparing the equality of items.
     * 
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+     * compare the items using that identifier.
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataView.java:86]:
```suggestion
```
There is nothing "compared" as to me that refers to sorting.

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.vaadin.flow.data.provider;
+
+import java.io.Serializable;
+
+/**
+ * Interface that defines methods for setting data.
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:22]:
```suggestion
 * A mixin interface for components that accept setting a data source
 * of the generic types {@link DataProvider} and {@link InMemoryDataProvider}.
 * The methods return a {@link DataView} which has the generic API
 * for getting information on the data.
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+ * This will return a {@link DataView}.
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:23]:
```suggestion
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ * @param <T>
+ *         data type
+ * @param <V>
+ *         DataView type
+ * @since
+ */
+public interface HasDataView<T, V extends DataView<T>> extends Serializable {
+
+    /**
+     * Set a generic data provider for the component to use.
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:34]:
```suggestion
     * Set a generic data provider for the component to use and
     * returns the base {@link DataView} that provides API to get
     * information on the data.
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+     * <p>
+     * It is suggested to use a more specific target from {@link
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:36]:
```suggestion
     * This method should be used only when the data provider type
     * is not either {@link ListDataProvider} or {@link LBackendDataProvider}.
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+     * HasListDataView} or {@link HasLazyDataView}
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:37]:
```suggestion
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * @param dataProvider
+     *         DataProvider instance to use
+     * @return DataView instance
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:41]:
```suggestion
     * @return DataView providing information on the data
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * Sets an InMemory data provider for the component to use.
+     * <p>
+     * Note! Using a {@link ListDataProvider} instead of a {@link
+     * InMemoryDataProvider} is recommended to get access to {@link
+     * ListDataView} API by using {@link HasListDataView#setDataSource(ListDataProvider)}.
+     *
+     * @param dataProvider
+     *         InMemoryDataProvider to use
+     * @return DataView instance
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasDataView.java:54]:
```suggestion
     * @return DataView providing information on the data
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java [hunk 13]
```diff
@@ -261,6 +266,51 @@ public void confirmUpdate(int updateId) {
...
+        }
+        return assumedSize;
+    }
+
+    /**
+     * Get the active keys and order on active data on the client.
+     *
+     * @return list of active client data as ordered keys
+     */
+    public List<String> getActiveKeyOrdering() {
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java:287]: Not used

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java [hunk 14]
```diff
@@ -261,6 +266,51 @@ public void confirmUpdate(int updateId) {
...
+        return Collections.unmodifiableList(activeKeyOrder);
+    }
+
+    /**
+     * Get the current client item range.
+     *
+     * @return range of items on client
+     */
+    public Range getRequestedRange() {
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java:296]: Not used

**Comment by pleku** [comment]
@mshabarov can you check trough the fixed I added ?

**Comment by mshabarov** [comment]
> @mshabarov can you check trough the fixed I added ?

Yes, in progress

**Comment by mshabarov** [comment]
@pleku your changes look good, i have no significant comments so far.

**Pull request merged**